examples/gsm8k-lora/gsm8k_lora.py

```diff
@@ -0,0 +1,221 @@
+from levanter.utils.hf_utils import num_cpus_used_by_tokenizer
+from levanter.utils.jax_utils import parameter_count
+from levanter.utils.py_utils import non_caching_cycle
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class TrainArgs:
+    optimizer: OptimizerConfig
+    trainer: TrainerConfig
+    lora: LoraConfig = LoraConfig()
+
+    max_tune_length: int = 2048  # maximum length of the input to the model during tuning
+    data: str = "tatsu-lab/alpaca"  # Path to the training data, or huggingface dataset name.
+    data_cache_dir: str = "cache/"  # Path to cache the tokenized data. can be gcs
+
+    input_key: str = "question"  # key in the dataset for the input
+    output_key: str = "answer"  # key in the dataset for the output
+    mask_inputs: bool = True  # if True, mask out the input and prompt for loss calculation
+
+    model_name_or_path: str = "meta-llama/Llama-2-7b-hf"
+    trust_remote_code: bool = False  # Trust remote code when loading from HuggingFace checkpoints.
+
+    model_cache_dir: Optional[str] = None  # Path to cache the model. must be local.
+
+    hf_save_path: Optional[str] = None  # Path to save the HuggingFace checkpoint.
+    hf_upload: Union[bool, str] = False  # Name of the HuggingFace repo to upload to (if any).
+    hf_save_steps: int = 1000  # How often to save the HuggingFace checkpoint.
+
+    # should we save merged (i.e. not peft) checkpoints?
+    merged_hf_save_path: Optional[str] = None  # path to save merged hf checkpoints
+    merged_hf_upload: Optional[str] = None
+
+
+class SupervisedDataset(Dataset[LmExample]):
+    def __init__(self, preproc_dataset, tokenizer, mask_inputs):
+        self.preproc_dataset = preproc_dataset
+        self.tokenizer = tokenizer
+        self.mask_inputs = mask_inputs
+
+    def __iter__(self):
+        for ex in self.preproc_dataset:
+            # annoyingly, pad expects things to be batched so we have to prepend a batch axis
+            ex = self.tokenizer.pad(
+                {k: np.expand_dims(v, 0) for k, v in ex.items()}, return_tensors="np", padding="max_length"
+            )
+            ex = {k: v[0] for k, v in ex.items()}
+            input_ids = hax.named(ex["input_ids"], "position")
+
+            # mask out padding and anything before the start of the target
+            Pos = input_ids.resolve_axis("position")
+            if self.mask_inputs:
+                loss_mask = hax.arange(Pos) >= ex["source_lens"]
+
+                # don't predict the padding
+                targets = hax.roll(input_ids, -1, Pos)
+                loss_mask = loss_mask & (targets != self.tokenizer.pad_token_id)
+            else:
+                loss_mask = 1 - hax.nn.one_hot(-1, Pos, dtype=jax.numpy.float32)
+
+            yield LmExample(input_ids, loss_mask)
+
+
+def mk_dataset(config: TrainArgs, tokenizer: transformers.PreTrainedTokenizerBase):
+    dataset = WrappedHFDataset("gsm8k", split="train", name="main")
+
+    def preprocess(batch):
```

Comment:
Not sure if I miss anything else. Is this data formatting the main place where it differs from the existing alpaca and alpaca_lora?

Reply:
yeah everything else is cosmetic

Reply:
i added the ability to choose whether to mask inputs for the loss calculation, i guess